src/core/chunked_stream.js

```diff
@@ -290,283 +276,274 @@ var ChunkedStreamManager = (function ChunkedStreamManagerClosure() {
-
-      var i, ii;
-      var chunksNeeded = Object.create(null);
-      this.chunksNeededByRequest[requestId] = chunksNeeded;
-      for (i = 0, ii = chunks.length; i < ii; i++) {
-        if (!this.stream.hasChunk(chunks[i])) {
-          chunksNeeded[chunks[i]] = true;
-        }
-      }
+      };
+      rangeReader.read().then(readChunk, reject);
+    });
+    promise.then((data) => {
+      if (this.aborted) {
+        return; // Ignoring any data after abort.
+      }
+      this.onReceiveData({ chunk: data, begin, });
+    });
+    // TODO check errors
+  }
 
-      if (isEmptyObj(chunksNeeded)) {
-        return Promise.resolve();
-      }
+  /**
+   * Get all the chunks that are not yet loaded and group them into
+   * contiguous ranges to load in as few requests as possible.
+   */
+  requestAllChunks() {
+    const missingChunks = this.stream.getMissingChunks();
+    this._requestChunks(missingChunks);
+    return this._loadedStreamCapability.promise;
+  }
 
-      var capability = createPromiseCapability();
-      this.promisesByRequest[requestId] = capability;
+  _requestChunks(chunks) {
+    const requestId = this.currRequestId++;
 
-      var chunksToRequest = [];
-      for (var chunk in chunksNeeded) {
-        chunk = chunk | 0;
-        if (!(chunk in this.requestsByChunk)) {
-          this.requestsByChunk[chunk] = [];
-          chunksToRequest.push(chunk);
-        }
-        this.requestsByChunk[chunk].push(requestId);
+    const chunksNeeded = Object.create(null);
+    this.chunksNeededByRequest[requestId] = chunksNeeded;
+    for (let i = 0, ii = chunks.length; i < ii; i++) {
+      if (!this.stream.hasChunk(chunks[i])) {
+        chunksNeeded[chunks[i]] = true;
       }
+    }
 
-      if (!chunksToRequest.length) {
-        return capability.promise;
-      }
+    if (isEmptyObj(chunksNeeded)) {
+      return Promise.resolve();
+    }
 
-      var groupedChunksToRequest = this.groupChunks(chunksToRequest);
+    const capability = createPromiseCapability();
+    this.promisesByRequest[requestId] = capability;
 
-      for (i = 0; i < groupedChunksToRequest.length; ++i) {
-        var groupedChunk = groupedChunksToRequest[i];
-        var begin = groupedChunk.beginChunk * this.chunkSize;
-        var end = Math.min(groupedChunk.endChunk * this.chunkSize, this.length);
-        this.sendRequest(begin, end);
+    const chunksToRequest = [];
+    for (let chunk in chunksNeeded) {
+      chunk = chunk | 0;
+      if (!(chunk in this.requestsByChunk)) {
+        this.requestsByChunk[chunk] = [];
+        chunksToRequest.push(chunk);
       }
+      this.requestsByChunk[chunk].push(requestId);
+    }
 
+    if (!chunksToRequest.length) {
       return capability.promise;
-    },
+    }
 
-    getStream: function ChunkedStreamManager_getStream() {
-      return this.stream;
-    },
+    const groupedChunksToRequest = this.groupChunks(chunksToRequest);
+    for (let i = 0; i < groupedChunksToRequest.length; ++i) {
```

Comment:
Could potentially use `for (const groupedChunked of groupedChunksToRequest) {` here instead.